python/perf-kernels/fused_moe/moe-gemm.py

```diff
@@ -0,0 +1,495 @@
+    kernel on a given batch size bs, the closest batch size in the grid should
+    be picked and the associated configuration chosen to invoke the kernel.
+    """
+    # First look up if an optimized configuration is available in the configs
+    # directory
+    json_file_name = get_config_file_name(dtype)
+
+    config_file_path = os.path.join(os.path.dirname(os.path.realpath(__file__)), "configs", json_file_name)
+    if os.path.exists(config_file_path):
+        with open(config_file_path) as f:
+            # If a configuration has been found, return it
+            return {key: val for key, val in json.load(f).items()}
+
+    # If no optimized configuration is available, we will use the default
+    # configuration
+    return None
+
+
+def get_default_config(
+    M: int,
+    E: int,
+    is_marlin: bool,
+) -> Dict[str, int]:
+    config = {'BLOCK_SIZE_M': 64, 'BLOCK_SIZE_N': 64, 'BLOCK_SIZE_K': 32, 'GROUP_SIZE_M': 8}
+    # A heuristic: fused marlin works faster with this config for small M
+    if M <= E or (is_marlin and M <= 32):
+        config = {'BLOCK_SIZE_M': 16, 'BLOCK_SIZE_N': 32, 'BLOCK_SIZE_K': 64, 'GROUP_SIZE_M': 1}
+    return config
+
+
+def try_get_optimal_moe_config(
+    E: int,
+    dtype: Optional[str],
+    M: int,
+    is_marlin: bool = False,
+):
+    configs = get_moe_configs(dtype)
+
+    if configs:
+        if configs:
+            if M < M_THRESHOLD_SMALL:
+                config = configs["small_M"]
+            elif M < M_THRESHOLD_MEDIUM:
+                config = configs["medium_M"]
+            else:
+                config = configs["large_M"]
+    else:
+        # Else use the default config
+        config = get_default_config(M, E, is_marlin)
+
+    return config
+
+
+def moe_gemm(a: torch.Tensor, b: torch.Tensor, c: torch.Tensor, topk_weights: torch.Tensor, topk_ids: torch.Tensor,
+             sorted_token_ids: torch.Tensor, expert_ids: torch.Tensor, num_tokens_post_padded: torch.Tensor,
+             config) -> None:
+    # TODO shard M dim
+    _, top_k = topk_ids.shape
+
+    EM = num_tokens_post_padded.item()
+    _, N, K = b.shape
+    grid = lambda META: (triton.cdiv(EM, META['BLOCK_SIZE_M']) * triton.cdiv(N, META['BLOCK_SIZE_N']), )
+
+    moe_gemm_kernel[grid](a, b, c, a.stride(0), a.stride(1), b.stride(0), b.stride(1), b.stride(2), c.stride(1),
+                          c.stride(2), top_k, topk_weights, sorted_token_ids, expert_ids, EM, N, K,
+                          MUL_ROUTED_WEIGHT=topk_weights is not None, **config)
+    return c
+
+
+def input_helper(M: int, N: int, K: int, top_k: int, E: int, routed_weight: bool, dtype):
+    a = torch.randn((M, K), dtype=dtype, device='cuda')
+    b = torch.randn((E, N, K), dtype=dtype, device='cuda')
+    c = torch.zeros((M, top_k, N), dtype=dtype, device='cuda')
+
+    values = torch.randn(M, E, device='cuda')
+
+    softmax_vals = torch.softmax(values, dim=1)
+    topk_weights, topk_ids = torch.topk(softmax_vals, k=top_k, dim=1)
```

Comment:
This is quite random. Does it represent how things work in real model?
I'm asking bc a while ago I saw different performance of the moe kernel between e2e and micro-benchmark.

Reply:
Could you elaborate? What is the shape, and datatype?

Reply:
I checked the issue again and found that the gap I saw was caused by that the ubench does not use softmax in the gating function. Since softmax is used here, I think we should be fine.